tensorstore/kvstore/s3/s3_request_builder.cc

```diff
@@ -0,0 +1,278 @@
+#include "absl/strings/escaping.h"
+#include "absl/strings/str_split.h"
+#include "absl/status/status.h"
+#include "tensorstore/kvstore/s3/s3_request_builder.h"
+#include "tensorstore/kvstore/s3/validate.h"
+#include "tensorstore/internal/uri_utils.h"
+#include "tensorstore/internal/path.h"
+#include "tensorstore/internal/digest/sha256.h"
+#include "tensorstore/util/result.h"
+#include "tensorstore/util/str_cat.h"
+
+using ::tensorstore::internal::AsciiSet;
+using ::tensorstore::internal::ParseGenericUri;
+using ::tensorstore::internal::ParsedGenericUri;
+using ::tensorstore::internal::SHA256Digester;
+using namespace ::tensorstore::internal_http;
+using ::tensorstore::internal_storage_s3::IsValidBucketName;
+
+#ifndef TENSORSTORE_INTERNAL_S3_LOG_AWS4
+#define TENSORSTORE_INTERNAL_S3_LOG_AWS4 0
+#endif
+
+namespace tensorstore {
+namespace internal_storage_s3 {
+
+
+namespace {
+
+// See description of function UriEncode at this URL
+// https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+constexpr AsciiSet kUriUnreservedChars{
+    "abcdefghijklmnopqrstuvwxyz"
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    "0123456789"
+    "-._~"};
+
+// NOTE: Only adds "/" to kUriUnreservedChars
+constexpr AsciiSet kUriKeyUnreservedChars{
+    "abcdefghijklmnopqrstuvwxyz"
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    "0123456789"
+    "/-._~"};
+
+
+/// Size of HMAC (size of SHA256 digest).
+constexpr static size_t kHmacSize = 32;
+
+void ComputeHmac(std::string_view key, std::string_view message, unsigned char (&hmac)[kHmacSize]) {
+    unsigned int md_len = kHmacSize;
+    // Computing HMAC should never fail.
+    ABSL_CHECK(HMAC(EVP_sha256(),
+                    reinterpret_cast<const unsigned char*>(key.data()),
+                    key.size(),
+                    reinterpret_cast<const unsigned char*>(message.data()),
+                    message.size(), hmac, &md_len) &&
+               md_len == kHmacSize);
+}
+
+void ComputeHmac(unsigned char (&key)[kHmacSize], std::string_view message, unsigned char (&hmac)[kHmacSize]){
+    unsigned int md_len = kHmacSize;
+    // Computing HMAC should never fail.
+    ABSL_CHECK(HMAC(EVP_sha256(), key, kHmacSize,
+                    reinterpret_cast<const unsigned char*>(message.data()),
+                    message.size(), hmac, &md_len) &&
+               md_len == kHmacSize);
+}
+
+
+} // namespace
+
+std::string UriEncode(std::string_view src) {
+  std::string dest;
+  internal::PercentEncodeReserved(src, dest, kUriUnreservedChars);
+  return dest;
+}
+
+std::string UriObjectKeyEncode(std::string_view src) {
+  std::string dest;
+  internal::PercentEncodeReserved(src, dest, kUriKeyUnreservedChars);
+  return dest;
+}
+
+S3RequestBuilder & S3RequestBuilder::AddHeader(std::string_view header) {
+  auto pos = header.find(':');
+  assert(pos != std::string::npos);
+
+  builder_.AddHeader(header);
+  auto key = absl::AsciiStrToLower(absl::StripAsciiWhitespace(header.substr(0, pos)));
```

Comment:
You could just normalize and sort the headers at the end, in `BuildRequest`, rather than do an insertion sort here.  Same for the query parameters.

Reply:
Just to clarify I take this to mean that:

1. `HttpRequest.headers` should be normalised in `BuildRequest`, as opposed to storing header pairs in `S3RequestBuilder.signed_headers_` which are then normalised in `BuildRequest`
2. Query Parameters should be extracted from the `HttpRequest.url` and normalised in `BuildRequest`, as opposed to storing query parameters in `S3RequestBuilder.query_params_` which are then normalised in `BuildRequest`

Reply:
1. Yes
2. No, I think storing a separate list of query parameters, as you do now, is reasonable, you just don't need to sort it until BuildRequest is called.

Reply:
done